### ISSUE TITLE
SelfAssertion for Truth incorrectly fires on AssertJ assertions

### ISSUE THREAD

**Issue by TimvdLippe** [google/error-prone#5131]
Given the following Mockito code:

```java
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

public class HashCodeAndEqualsSafeSetTest {

    @Rule public MockitoRule r = MockitoJUnit.rule();
    @Mock private UnmockableHashCodeAndEquals mock1;

    @Test
    public void isEqualToItself() {
        HashCodeAndEqualsSafeSet set = HashCodeAndEqualsSafeSet.of(mock1);
        assertThat(set).isEqualTo(set);
    }
}
```

Here we test that mocks are correctly implementing the equals contract. For it, we use AssertJ, which according to its docs we should use this exact pattern: https://assertj.github.io/doc/#assertj-core-incorrect-usage

However, https://errorprone.info/bugpattern/SelfAssertion fires, which is a rule for Truth instead of AssertJ.

This started failing in https://github.com/mockito/mockito/pull/3674 and I suspect this is https://github.com/google/error-prone/commit/962f5d1e17cd9f2583ac093c6ab955b2335b7cf1

It looks like this was exactly the intention, yet AssertJ actively promotes this pattern in its docs. Can we therefore exempt AssertJ from this check?

**Comment by graememorgan** [comment]
Could you point me to where AssertJ suggests this? I'm struggling to see where the linked doc mentions using AssertJ to test the validity of an `equals` implementation.

Our recommendation would usually be `EqualsTester`.

**Comment by TimvdLippe** [comment]
Yes, see the first link in my post

**Comment by cpovirk** [comment]
Can you give us a term to search for? Chrome says the page has 252 references to "isEqualTo" :) I skimmed them, and the only self-assertions I noticed were comparing `123` to itself and `"a"` (IIRC) to itself, neither of which looks to be a test of an `equals` implementation. 

(I think the question of what to do about AssertJ will be interesting even assuming that it does recommend this pattern, since it's easy to pass the same value twice accidentally.)

**Comment by TimvdLippe** [comment]
Apologies, the docs aren't super clear about this indeed. I found https://github.com/assertj/assertj/issues/1949 and corresponding PR https://github.com/assertj/assertj/pull/2114 that implements all checks for the equals contract.

**Comment by scordio** [comment]
Hi @TimvdLippe, I'm not aware that we actively promote `isEqualTo()` as a way to test the correctness of a custom `equals`/`hashCode` implementation. If you got that impression from the documentation, please let me know where, and we'll make it more explicit.

As mentioned in assertj/assertj#1949:

> AssertJ `isEqualTo` is not designed to provide full coverage for custom equals implementations. We usually suggest EqualsVerifier to achieve this.

assertj/assertj#2114 could have given the impression that we took a different decision. In reality, we only wanted to ensure that the most common use cases are better covered, leaving fewer surprises for users. However, our implementation is far less comprehensive than the work dedicated libraries like Guava's EqualsTester or [EqualsVerifier](https://jqno.nl/equalsverifier/) do behind the scenes, and we have no intention of targeting such a scope.

**Comment by TimvdLippe** [comment]
> If you got that impression from the documentation, please let me know where, and we'll make it more explicit.

Okay, that's new information to me, but understandable that AssertJ doesn't implement it after all. Essentially, when I read the api `.isEqualTo`, I would expect it to perform equals checks. E.g. we need to assert that our mocks correctly pass on the `equals` to the underlying class. If we weren't doing that, then two mocks of the same type would not be equal. And that would be observable behavior in production code, which is not what mocks are supposed to do.

Therefore, based on `foo.isEqualTo(bar)`, I expect that to call `foo.equals(bar)` and use that to throw the correct assertion. Maybe that's incorrect, but I don't see this mentioned on https://assertj.github.io/doc/ Then I look at https://assertj.github.io/doc/#assertj-core-incorrect-usage and I see that it promotes `isEqualTo` with `actual` and `expected`. Given that they are two different values, I expect `==` to not be used in AssertJ (since that would fail), thus I assume that AssertJ calls `actual.isEqualTo(expected)` to run its assertions.

Is that understanding correct?

**Comment by TimvdLippe** [comment]
For what is worth, the `SelfAssertion` is a good check. Since often it is a typo, so it catches mistakes by developers. Maybe our disabling of the rule for our specific case is a good compromise, since we are an outlier. The only thing that I am questioning is the fact that the documentation of `SelfAssertion` claims that assertion libraries use reference equality to compare objects. For libraries that only use reference equality, yes these are bugs. However, if an assertion library calls `foo.equals(bar)`, then I don't think the assertion should fire. I don't have a definitive list of assertion libraries that do so, nor which ones promise to do so.

**Comment by graememorgan** [comment]
I think I agree with this line in the documentation:

> To test the implementation of an `equals` method, use Guava's EqualsTester, or explicitly call `equals` as part of the test.

If one specifically wishes to confirm that calling `equals` returns `true`, then `assertThat(a.equals(b)).isTrue()` is a sensible assertion.

**Comment by cpovirk** [comment]
https://github.com/google/error-prone/pull/5150 improves the documentation.

Someday, I will [figure out a good way to express why `isEqualTo` is not for testing `equals`](https://github.com/google/truth/issues/268) :) Arguably we should have called it "`is`" in the first place. (But it turns out that that would have gone poorly when we starting acquiring Kotlin users, since "`is`" is a Kotlin keyword.)

**Comment by TimvdLippe** [comment]
Thanks for the additional clarification in the documentation! That does help.

I think my fundamental misunderstanding of the API is the existence of `isSameAs`. See this popular StackOverflow question that perpetuates the same notion: https://stackoverflow.com/questions/76638706/idiomatic-way-to-force-assertj-to-compare-by-identity

In fact, while trying to lookup the docs of AssertJ, it does have `assertThat("abc").isEqualTo("abc");` as its example: https://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/Assert.html#isEqualTo(java.lang.Object) Similarly, https://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/Assert.html#isSameAs(java.lang.Object) explicitly mentions it uses `==` comparison.

Basically, the fact that `isSameAs` exists makes me believe that `isEqualTo` does something different. Why would you otherwise have two equivalent methods in your API? And the former I would reasonably believe to use `==` based on its name and thus the latter doesn't use it.

Anyhow, in Mockito, I will update the code to be `assertThat(set.equals(set)).isTrue()` to make it explicit according to the ErrorProne documentation. I will close this ticket here and others can use it as documentation from their search results when they stumble upon it.

**Comment by scordio** [comment]
> Basically, the fact that `isSameAs` exists makes me believe that `isEqualTo` does something different.

@TimvdLippe They absolutely do something different!

As you already found out, `isSameAs` is for reference equality, while `isEqualTo` is for "value" equality, _eventually_ relying on the underlying `equals` implementation.

To summarize:
* Is `isEqualTo` using `equals` under the hood? [Yes](https://github.com/assertj/assertj/blob/bc67ddb83ed9798731a3a1dfd6c572d83954e01e/assertj-core/src/main/java/org/assertj/core/api/comparisonstrategy/StandardComparisonStrategy.java#L106).
* Is `isEqualTo` the way to test that a class `equals` implementation has been done correctly? Perhaps, but we provide no guarantee about it, and we suggest using dedicated libraries like EqualsTester or EqualsVerifier instead.

I created assertj/assertj#3865 to improve our documentation and make the latter clearer. Feel free to chime in if you'd like to discuss further details or pain points 🙂 and thank you for the input you've already provided in this thread!

**Comment by TimvdLippe** [comment]
Thanks for the clarification. In our case, we were only interested in the fact that it would call `equals` under the hood. So that reinforces the point that our usage was legit. Still, I understand why this ErrorProne rule was written as-is and that we are the outlier. I hope all of this information is useful to folks in the future who might have similar expectations.
